[pytorch] move tracing logic to a separate dispatch backend

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38467 [pytorch] move tracing logic to a separate dispatch backend**

This PR moves tracing logic out of the generated VariableType kernels, to associate it with a new dedicated dispatch key Tracer.
It also toggles the dispatch key set at various places to keep the semantics unchanged - see the inline [Tracing Mode Switches] note.

Sample generated code:
```
Tensor & __ilshift___Tensor(Tensor & self, const Tensor & other) {
  #if !defined(PYTORCH_DISABLE_TRACING)
  torch::jit::Node* node = nullptr;
  std::shared_ptr<jit::tracer::TracingState> tracer_state;
  if (jit::tracer::isTracing()) {
    tracer_state = jit::tracer::getTracingState();
    at::Symbol op_name;
    op_name = jit::Symbol::fromQualString("aten::__ilshift__");
    node = tracer_state->graph->create(op_name, /*num_outputs=*/0);
    jit::tracer::recordSourceLocation(node);
    jit::tracer::addInputs(node, "self", self);
    jit::tracer::addInputs(node, "other", other);
    tracer_state->graph->insertNode(node);

    jit::tracer::setTracingState(nullptr);
  }
  #endif
  static auto op = c10::Dispatcher::singleton().findSchemaOrThrow("aten::__ilshift__", "Tensor");
  c10::Dispatcher::singleton().redispatch<Tensor &, Tensor &, const Tensor &>(op, c10::DispatchKey::Tracer, self, other);
  #if !defined(PYTORCH_DISABLE_TRACING)
  if (tracer_state) {
    jit::tracer::setTracingState(std::move(tracer_state));
    jit::tracer::addOutput(node, self);
  }
  #endif
  return self;
}
```



Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21570684/)!

Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684)

ljk53 committed:
[WIP][pytorch] move tracing logic from VariableType to TraceType



Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21570684/)!

[ghstack-poisoned]

ljk53 committed:
Update on "[WIP][pytorch] move tracing logic from VariableType to TraceType"


Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21570684/)!

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] move tracing logic from VariableType to TraceType"


Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21570684/)!

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] move tracing logic from VariableType to TraceType"


Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21570684/)!

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] move tracing logic to a separate dispatch backend (part 1: codegen only)"






Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21570684/)!

Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684)

[ghstack-poisoned]

ljk53 commented:
This PR can be reviewed - it only generates the new TraceType_*.cpp but doesn't build them yet.
The integration part is in the next PR.

ljk53 committed:
Update on "[pytorch] move tracing logic to a separate dispatch backend (part 1: codegen only)"






Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21570684/)!

Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684)

[ghstack-poisoned]

ezyang commented:
This looks reasonable, though I only skimmed the codegen bits. You'll probably want to land this all in one go with the second part right?

bhosmer commented on tools/autograd/gen_variable_type.py:
```diff
@@ -350,6 +350,18 @@
+
+# TraceType templates
+TRACE_DISPATCH_FIND_OP = CodeTemplate("""\
+static auto op = c10::Dispatcher::singleton().findSchema({"aten::${operator_name}", "${overload_name}"});
```
Can we factor things to eliminate the duplication between these `TRACE_DISPATCH_FIND_OP` and `TRACE_DISPATCH_CALL_OP` and `PROFILE_DISPATCH_UNBOXED` above? I know that takes you outside the new functionality but it seems like it's worth doing.

Oh also! I think you can replace the `findSchema()/TORCH_INTERNAL_ASSERT()` with `findSchemaOrThrow()` - see e.g. [here](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/function_wrapper.py#L136)

ljk53 replied:
Switched to `findSchemaOrThrow()` and merged `TRACE_DISPATCH_FIND_OP` & `TRACE_DISPATCH_CALL_OP`. I didn't touch `PROFILE_DISPATCH_UNBOXED` because I was told @ilia-cher  is going to remove it (calling profiler from dispatcher instead).

`TRACE_DISPATCH_FIND_OP` / `TRACE_DISPATCH_CALL_OP` were initially separated because `auto result = ` might be inserted before `TRACE_DISPATCH_CALL_OP` when it doesn't return void / mutating self. Now I pass "auto result = " into the CodeTemplate - however, I figured it unconditionally strips the trailing whitespace in the value [here](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/code_template.py#L44), so it will look like "auto result =c10::Dispatcher::singleton()..." - correct but ugly :(

bhosmer commented on tools/autograd/gen_variable_type.py:
```diff
@@ -988,40 +1085,18 @@ def emit_call(env):
                     base_type_call=base_type_call)
 
-                call += wrap_output(tie_return_values(), 'tmp')
+                call += wrap_output(tie_return_values, 'tmp')
```
I don't know if this is too fussy, but I think the code would be easier to follow if `tie_return_values` was now passed into `emit_call()` explicitly. (Actually FWIW I think having `format_return_variables()` gather those three bits of codegen together might be making the logistics a bit higher-friction than if they were still independent, though ofc factoring to make it/them callable from both tracing and variable is totally the right thing to do, and I see why you brought them together.)

Totally up to you if you want to change any of this - basically just nits.

ljk53 replied:
Passed in tie_return_values explicitly.

Yeah - I grouped them together to share these util methods...
```
    arguments = declaration['arguments']
    returns = declaration['returns']
    func = declaration['derivative']
    name = declaration['name']
    inplace = declaration['inplace']
    is_out_fn = name.endswith('_out')
    modifies_arguments = inplace or is_out_fn
    returns_void = len(returns) == 0
```

bhosmer commented on tools/autograd/templates/TraceType.cpp:
```diff
@@ -0,0 +1,31 @@
+
+#include "torch/csrc/autograd/function.h"
+
+// ${generated_comment}
```
Might be worth adding a (non-generated) comment that briefly describes what these kernels do?

ljk53 replied:
Added a link to `torch/csrc/jit/OVERVIEW.md`.

bhosmer reviewed (approved):
Looks good! Couple minor comments inline.

ezyang commented:
Can you please post some examples of the generated code?

ljk53 committed:
Update on "[pytorch] move tracing logic to a separate dispatch backend (part 1: codegen only)"


This PR is the first step to move tracing logic into a separate dispatch backend.

Sample generated code:
```
Tensor & __ilshift___Tensor(Tensor & self, const Tensor & other) {
  #if !defined(PYTORCH_DISABLE_TRACING)
  torch::jit::Node* node = nullptr;
  std::shared_ptr<jit::tracer::TracingState> tracer_state;
  if (jit::tracer::isTracing()) {
    tracer_state = jit::tracer::getTracingState();
    at::Symbol op_name;
    op_name = jit::Symbol::fromQualString("aten::__ilshift__");
    node = tracer_state->graph->create(op_name, /*num_outputs=*/0);
    jit::tracer::recordSourceLocation(node);
    jit::tracer::addInputs(node, "self", self);
    jit::tracer::addInputs(node, "other", other);
    tracer_state->graph->insertNode(node);

    jit::tracer::setTracingState(nullptr);
  }
  #endif
  static auto op = c10::Dispatcher::singleton().findSchema({"aten::__ilshift__", "Tensor"});
  TORCH_INTERNAL_ASSERT(op);
  c10::Dispatcher::singleton().redispatch<Tensor &, Tensor &, const Tensor &>(*op, c10::DispatchKey::Tracer, self, other);
  #if !defined(PYTORCH_DISABLE_TRACING)
  if (tracer_state) {
    jit::tracer::setTracingState(std::move(tracer_state));
    jit::tracer::addOutput(node, self);
  }
  #endif
  return self;
}
```



Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21570684/)!

Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] move tracing logic to a separate dispatch backend (part 1: codegen only)"


This PR is the first step to move tracing logic into a separate dispatch backend.

Sample generated code:
```
Tensor & __ilshift___Tensor(Tensor & self, const Tensor & other) {
  #if !defined(PYTORCH_DISABLE_TRACING)
  torch::jit::Node* node = nullptr;
  std::shared_ptr<jit::tracer::TracingState> tracer_state;
  if (jit::tracer::isTracing()) {
    tracer_state = jit::tracer::getTracingState();
    at::Symbol op_name;
    op_name = jit::Symbol::fromQualString("aten::__ilshift__");
    node = tracer_state->graph->create(op_name, /*num_outputs=*/0);
    jit::tracer::recordSourceLocation(node);
    jit::tracer::addInputs(node, "self", self);
    jit::tracer::addInputs(node, "other", other);
    tracer_state->graph->insertNode(node);

    jit::tracer::setTracingState(nullptr);
  }
  #endif
  static auto op = c10::Dispatcher::singleton().findSchemaOrThrow("aten::__ilshift__", "Tensor");
  c10::Dispatcher::singleton().redispatch<Tensor &, Tensor &, const Tensor &>(op, c10::DispatchKey::Tracer, self, other);
  #if !defined(PYTORCH_DISABLE_TRACING)
  if (tracer_state) {
    jit::tracer::setTracingState(std::move(tracer_state));
    jit::tracer::addOutput(node, self);
  }
  #endif
  return self;
}
```



Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21570684/)!

Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684)

[ghstack-poisoned]

ljk53 commented:
> Can you please post some examples of the generated code?

Added to the PR description.

ljk53 committed:
Update on "[pytorch] move tracing logic to a separate dispatch backend (part 1: codegen only)"


This PR is the first step to move tracing logic into a separate dispatch backend.

Sample generated code:
```
Tensor & __ilshift___Tensor(Tensor & self, const Tensor & other) {
  #if !defined(PYTORCH_DISABLE_TRACING)
  torch::jit::Node* node = nullptr;
  std::shared_ptr<jit::tracer::TracingState> tracer_state;
  if (jit::tracer::isTracing()) {
    tracer_state = jit::tracer::getTracingState();
    at::Symbol op_name;
    op_name = jit::Symbol::fromQualString("aten::__ilshift__");
    node = tracer_state->graph->create(op_name, /*num_outputs=*/0);
    jit::tracer::recordSourceLocation(node);
    jit::tracer::addInputs(node, "self", self);
    jit::tracer::addInputs(node, "other", other);
    tracer_state->graph->insertNode(node);

    jit::tracer::setTracingState(nullptr);
  }
  #endif
  static auto op = c10::Dispatcher::singleton().findSchemaOrThrow("aten::__ilshift__", "Tensor");
  c10::Dispatcher::singleton().redispatch<Tensor &, Tensor &, const Tensor &>(op, c10::DispatchKey::Tracer, self, other);
  #if !defined(PYTORCH_DISABLE_TRACING)
  if (tracer_state) {
    jit::tracer::setTracingState(std::move(tracer_state));
    jit::tracer::addOutput(node, self);
  }
  #endif
  return self;
}
```



Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21570684/)!

Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] move tracing logic to a separate dispatch backend (part 1: codegen only)"


This PR is the first step to move tracing logic into a separate dispatch backend.

Sample generated code:
```
Tensor & __ilshift___Tensor(Tensor & self, const Tensor & other) {
  #if !defined(PYTORCH_DISABLE_TRACING)
  torch::jit::Node* node = nullptr;
  std::shared_ptr<jit::tracer::TracingState> tracer_state;
  if (jit::tracer::isTracing()) {
    tracer_state = jit::tracer::getTracingState();
    at::Symbol op_name;
    op_name = jit::Symbol::fromQualString("aten::__ilshift__");
    node = tracer_state->graph->create(op_name, /*num_outputs=*/0);
    jit::tracer::recordSourceLocation(node);
    jit::tracer::addInputs(node, "self", self);
    jit::tracer::addInputs(node, "other", other);
    tracer_state->graph->insertNode(node);

    jit::tracer::setTracingState(nullptr);
  }
  #endif
  static auto op = c10::Dispatcher::singleton().findSchemaOrThrow("aten::__ilshift__", "Tensor");
  c10::Dispatcher::singleton().redispatch<Tensor &, Tensor &, const Tensor &>(op, c10::DispatchKey::Tracer, self, other);
  #if !defined(PYTORCH_DISABLE_TRACING)
  if (tracer_state) {
    jit::tracer::setTracingState(std::move(tracer_state));
    jit::tracer::addOutput(node, self);
  }
  #endif
  return self;
}
```



Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21570684/)!

Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684)

[ghstack-poisoned]

ezyang commented:
A few notes from looking at the generated code:
* I think the `jit::tracer::isTracing()` is now unnecessary. Because we should just bypass the tracing key entirely if tracing is disabled, so you don't need to put these in a conditional. If getting this PR to pass all CI was challenging you can save this refactor for later, but we should definitely get rid of it at some point, and it may be easier to do so now while you have the patient open.
* Similarly, `PYTORCH_DISABLE_TRACING` also seems like it is in the wrong place now. @iseeyuan, correct me if I'm wrong, but this toggle is to get rid of the tracing code for federated learning on mobile; to me, this suggests that we should just avoid registering tracing wrappers *entirely* when we have this flag turned on.

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -350,6 +350,14 @@
+
+# TraceType templates
+TRACE_DISPATCH_UNBOXED = CodeTemplate("""\
+static auto op = c10::Dispatcher::singleton().findSchemaOrThrow("aten::${operator_name}", "${overload_name}");
```
No action needed for this PR, but we really need to come up with a better idiom for redispatching in these situations.

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -350,6 +350,14 @@
+# TraceType templates
+TRACE_DISPATCH_UNBOXED = CodeTemplate("""\
+static auto op = c10::Dispatcher::singleton().findSchemaOrThrow("aten::${operator_name}", "${overload_name}");
+${assign_return_values}c10::Dispatcher::singleton().redispatch<${ret_and_arg_types}>(${trace_dispatch_args});
```
Actually, I think `redispatch` may be the wrong thing to use in this case. The idea of redispatch is it lets you bypass the current dispatch key, without actually adding it to the exclusion set. But in the case of tracing, you always want to exclude tracing before continuing on (the same way variable gets excluded before you redispatch). So you really just want to do a "normal" redispatch, in the same way that VariableType wrappers did it.

ljk53 replied:
VariableType does so via `AutoNonVariableTypeMode` + calling `at::xxx()` again - for Tracing is a combination of `setTracingState(null)` (for other ops being called) and redispatch (for root op itself here).

Do you suggest using NoTracerDispatchMode here as well and change to regular `op.call` call? (note that I won't remove `setTracingState(null)` immediately in this diff because they might still be needed for factory methods.

ljk53 replied:
@ezyang I (speculatively) changed to NoTracerDispatchMode guard + call() to replace redispatch(). It's not the "same way that VariableType wrappers did it", though - because VariableType wrapper generator has logic to make direct TypeDefault::xxx() call for certain cases, which I cannot do because I should still redispatch to VariableType kernels first.

ezyang commented:
Also, I'm expecting to see some manual refactor on VariableTypeManual.cpp but I do not see it in this diff or the next one. The manual VariableType has tracing code in it that also has to be extracted, otherwise you'll have some problems.

ezyang commented:
Having done a closer review, I'm going to reiterate a point earlier: I think it would be much better if we squash this diff with the next one, to make sure we are actually exercising the code paths in question. That takes precedence over making things easier for us to review.

iseeyuan commented:
> A few notes from looking at the generated code:
> 
> * I think the `jit::tracer::isTracing()` is now unnecessary. Because we should just bypass the tracing key entirely if tracing is disabled, so you don't need to put these in a conditional. If getting this PR to pass all CI was challenging you can save this refactor for later, but we should definitely get rid of it at some point, and it may be easier to do so now while you have the patient open.
> * Similarly, `PYTORCH_DISABLE_TRACING` also seems like it is in the wrong place now. @iseeyuan, correct me if I'm wrong, but this toggle is to get rid of the tracing code for federated learning on mobile; to me, this suggests that we should just avoid registering tracing wrappers _entirely_ when we have this flag turned on.

Yes PYTORCH_DISABLE_TRACING and jit::tracer::isTracing() are used for FL in mobile. There's plan to use lite interpreter for FL and the testing binary code, lite_trainer is already there. It cannot compile with the tracing code. "we should just bypass the tracing key entirely if tracing is disabled", I assume it's done in code-gen stage?

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -592,19 +646,36 @@ def gen_variable_type_shard(out, aten_declarations, template_path, suffix, heade
                 declaration, class_type='ProfiledType'))
 
+        # Emit TraceType code
+        if not declaration['manual_kernel_registration']:
```
cc @smessmer I want to reiterate that it was a mistake to tie `manual_kernel_registration` to control BOTH whether or not we generate registrations for CPUType/TypeDefault, and whether or not we generate Variable/Profiling/Tracing wrappers here. I *absolutely* might want to do a manual registration on CPU/CUDA side, but still get tracing automatically generated for me.

Another thing that I find a bit perplexing is that the logic here doesn't line up with the profiling logic. Profiling appears to be generated unconditionally, even if we specified manual kernel registration. What's going on there?

ljk53 replied:
In my initial revision I followed the profiling logic - generating unconditionally, but later I found that some manual registered ops do NOT have tracing logic in it, e.g.: https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/VariableTypeManual.cpp#L122; generating a tracing kernel for them unconditionally changes semantics.

I think for Profiling the expectation is a little bit relaxed - it doesn't have to be 100% aligned with old behavior (and it doesn't for various corner cases I handled in next diff).

I think @ilia-cher had plan to move profiling into dispatcher? so I didn't change it to avoid conflict.

If not `manual_kernel_registration` then we will need some other blacklist mechanism to skip these ops if we want to keep them from tracing, or it does not matter to trace them?

ezyang replied:
OK, this answers my second question satisfactorily.  I agree as you've written it is semantics preserving and you should keep it as is. (I think my first point still stands but that's a matter for another day.)

ezyang commented on tools/autograd/gen_variable_type.py:
```diff
@@ -510,6 +518,48 @@ def format_trace(declaration):
 
 
+# Methods shared by TraceType and VariableType to handle return variable declaration, tie and tuple.
+def format_return_variables(declaration):
```
It looks like this was refactored from https://github.com/pytorch/pytorch/pull/38467/files#diff-bbfdc6df0a634ed4c7442fdc4129499dL1003 ; was there any substantive change in this refactor?

ljk53 replied:
No. BTW, you can review this on D21570684 - in Phabricator moved lines are highlighted with orange bar after line number :)

ljk53 commented:
> Also, I'm expecting to see some manual refactor on VariableTypeManual.cpp but I do not see it in this diff or the next one. The manual VariableType has tracing code in it that also has to be extracted, otherwise you'll have some problems.

Yes I'm aware of VariableTypeManual as I commented above - I plan to work on them next to fully decouple TraceType and VariableType - but I think the current two diffs alone do not break existing behavior, or I missed some cases?

ljk53 commented:
> A few notes from looking at the generated code:
> 
> * I think the `jit::tracer::isTracing()` is now unnecessary. Because we should just bypass the tracing key entirely if tracing is disabled, so you don't need to put these in a conditional. If getting this PR to pass all CI was challenging you can save this refactor for later, but we should definitely get rid of it at some point, and it may be easier to do so now while you have the patient open. EDIT: Actually, this PR is probably not testing this codepath in any substantive manner, so....
> * Similarly, `PYTORCH_DISABLE_TRACING` also seems like it is in the wrong place now. @iseeyuan, correct me if I'm wrong, but this toggle is to get rid of the tracing code for federated learning on mobile; to me, this suggests that we should just avoid registering tracing wrappers _entirely_ when we have this flag turned on.

I 100% agree that we should remove these two switches from generated code - I didn't do so on this diff yet because I wasn't sure how much additional changes it will bring in.

For `PYTORCH_DISABLE_TRACING` it might involve changing some internal buck/code for lite-interp build config - I know @iseeyuan is also touching those files recently so I'll wait to avoid merge conflict.

I think removing `jit::tracer::isTracing()` depends on moving out tracing logics from gen_variable_factories.py, right?: https://github.com/pytorch/pytorch/blob/master/tools/autograd/gen_variable_factories.py#L78
I plan to dig into it next, too.

ezyang commented:
> I think the current two diffs alone do not break existing behavior, or I missed some cases? 

I think you are probably right, but there is an annoying new failure mode in this new code which is if you (somehow) disable variable type but have tracing turned on, then all of the manual variable code will stop tracing. But I don't think you're likely to end up in this universe, unless we're trying to execute on the "make variable dispatch key aligned with requires grad". So yes, I guess it's fine to do later, assuming we actually do it later :)

> I think removing jit::tracer::isTracing() depends on moving out tracing logics from gen_variable_factories.py

and the manual variable type stuff too, probably.

ljk53 committed:
Update on "[pytorch] move tracing logic to a separate dispatch backend (part 1: codegen only)"


This PR is the first step to move tracing logic into a separate dispatch backend.

Sample generated code:
```
Tensor & __ilshift___Tensor(Tensor & self, const Tensor & other) {
  #if !defined(PYTORCH_DISABLE_TRACING)
  torch::jit::Node* node = nullptr;
  std::shared_ptr<jit::tracer::TracingState> tracer_state;
  if (jit::tracer::isTracing()) {
    tracer_state = jit::tracer::getTracingState();
    at::Symbol op_name;
    op_name = jit::Symbol::fromQualString("aten::__ilshift__");
    node = tracer_state->graph->create(op_name, /*num_outputs=*/0);
    jit::tracer::recordSourceLocation(node);
    jit::tracer::addInputs(node, "self", self);
    jit::tracer::addInputs(node, "other", other);
    tracer_state->graph->insertNode(node);

    jit::tracer::setTracingState(nullptr);
  }
  #endif
  static auto op = c10::Dispatcher::singleton().findSchemaOrThrow("aten::__ilshift__", "Tensor");
  c10::Dispatcher::singleton().redispatch<Tensor &, Tensor &, const Tensor &>(op, c10::DispatchKey::Tracer, self, other);
  #if !defined(PYTORCH_DISABLE_TRACING)
  if (tracer_state) {
    jit::tracer::setTracingState(std::move(tracer_state));
    jit::tracer::addOutput(node, self);
  }
  #endif
  return self;
}
```



Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21570684/)!

Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] move tracing logic to a separate dispatch backend"


This PR moves tracing logic out of the generated VariableType kernels, to associate it with a new dedicated dispatch key Tracer.
It also toggles the dispatch key set at various places to keep the semantics unchanged - see the inline [Tracing Mode Switches] note.

Sample generated code:
```
Tensor & __ilshift___Tensor(Tensor & self, const Tensor & other) {
  #if !defined(PYTORCH_DISABLE_TRACING)
  torch::jit::Node* node = nullptr;
  std::shared_ptr<jit::tracer::TracingState> tracer_state;
  if (jit::tracer::isTracing()) {
    tracer_state = jit::tracer::getTracingState();
    at::Symbol op_name;
    op_name = jit::Symbol::fromQualString("aten::__ilshift__");
    node = tracer_state->graph->create(op_name, /*num_outputs=*/0);
    jit::tracer::recordSourceLocation(node);
    jit::tracer::addInputs(node, "self", self);
    jit::tracer::addInputs(node, "other", other);
    tracer_state->graph->insertNode(node);

    jit::tracer::setTracingState(nullptr);
  }
  #endif
  static auto op = c10::Dispatcher::singleton().findSchemaOrThrow("aten::__ilshift__", "Tensor");
  c10::Dispatcher::singleton().redispatch<Tensor &, Tensor &, const Tensor &>(op, c10::DispatchKey::Tracer, self, other);
  #if !defined(PYTORCH_DISABLE_TRACING)
  if (tracer_state) {
    jit::tracer::setTracingState(std::move(tracer_state));
    jit::tracer::addOutput(node, self);
  }
  #endif
  return self;
}
```



Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21570684/)!

Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684)

[ghstack-poisoned]

ljk53 commented:
Merged integration part from #38502

ljk53 commented:
> I think you are probably right, but there is an annoying new failure mode in this new code which is if you (somehow) disable variable type but have tracing turned on, then all of the manual variable code will stop tracing.
Yes - to independently disable VariableType we need fix ManualVariableType which is what I said "plan to work on them next to fully decouple TraceType and VariableType" :)

ljk53 commented on tools/autograd/templates/variable_factories.h:
```diff
@@ -60,7 +61,7 @@ inline at::Tensor from_blob(
     const Deleter& deleter,
     const at::TensorOptions& options = at::TensorOptions()) {
   at::Tensor tensor = ([&]() {
-    at::AutoNonVariableTypeMode non_var_type_mode(true);
```
@ezyang I deleted most of `AutoNonVariableTypeMode` according to your comments on #38502, but I didn't ask there: what's the reason we had `AutoNonVariableTypeMode` at these places in the first place? was it solely to skip tracing logic? would rest part of VariableType (e.g.: things inside `if (compute_requires_grad...)` branch) have any undesired side effect for other ops being called? Are we making decision based on the fact that at::from_blob / at::empty / etc won't involve autograd logic at all?

ezyang replied:
Solely to skip tracing. This is something you have to work out by looking what gets called inside.

Actually, it was a bit *wrong* to skip variable mode, for example, I bet that `from_blob` when passed `requires_grad=True` didn't actually respect it. Your change here fixes that!

ljk53 replied:
Hmm, seems many CI failed because of removing AutoNonVariableTypeMode guard:
```
RuntimeError: Cannot insert a Tensor that requires grad as a constant. Consider making it a parameter or input, or detaching the gradient
```
I'll need figure out which one(s) of them are still necessary :'(

ljk53 committed:
Update on "[pytorch] move tracing logic to a separate dispatch backend"


This PR moves tracing logic out of the generated VariableType kernels, to associate it with a new dedicated dispatch key Tracer.
It also toggles the dispatch key set at various places to keep the semantics unchanged - see the inline [Tracing Mode Switches] note.

Sample generated code:
```
Tensor & __ilshift___Tensor(Tensor & self, const Tensor & other) {
  #if !defined(PYTORCH_DISABLE_TRACING)
  torch::jit::Node* node = nullptr;
  std::shared_ptr<jit::tracer::TracingState> tracer_state;
  if (jit::tracer::isTracing()) {
    tracer_state = jit::tracer::getTracingState();
    at::Symbol op_name;
    op_name = jit::Symbol::fromQualString("aten::__ilshift__");
    node = tracer_state->graph->create(op_name, /*num_outputs=*/0);
    jit::tracer::recordSourceLocation(node);
    jit::tracer::addInputs(node, "self", self);
    jit::tracer::addInputs(node, "other", other);
    tracer_state->graph->insertNode(node);

    jit::tracer::setTracingState(nullptr);
  }
  #endif
  static auto op = c10::Dispatcher::singleton().findSchemaOrThrow("aten::__ilshift__", "Tensor");
  {
    at::tracer::impl::NoTracerDispatchMode tracer_guard;
    c10::Dispatcher::singleton().call<Tensor &, Tensor &, const Tensor &>(op, self, other);
  }
  #if !defined(PYTORCH_DISABLE_TRACING)
  if (tracer_state) {
    jit::tracer::setTracingState(std::move(tracer_state));
    jit::tracer::addOutput(node, self);
  }
  #endif
  return self;
}
```



Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21570684/)!

Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] move tracing logic to a separate dispatch backend"


This PR moves tracing logic out of the generated VariableType kernels, to associate it with a new dedicated dispatch key Tracer.
It also toggles the dispatch key set at various places to keep the semantics unchanged - see the inline [Tracing Mode Switches] note.

Sample generated code:
```
Tensor & __ilshift___Tensor(Tensor & self, const Tensor & other) {
  #if !defined(PYTORCH_DISABLE_TRACING)
  torch::jit::Node* node = nullptr;
  std::shared_ptr<jit::tracer::TracingState> tracer_state;
  if (jit::tracer::isTracing()) {
    tracer_state = jit::tracer::getTracingState();
    at::Symbol op_name;
    op_name = jit::Symbol::fromQualString("aten::__ilshift__");
    node = tracer_state->graph->create(op_name, /*num_outputs=*/0);
    jit::tracer::recordSourceLocation(node);
    jit::tracer::addInputs(node, "self", self);
    jit::tracer::addInputs(node, "other", other);
    tracer_state->graph->insertNode(node);

    jit::tracer::setTracingState(nullptr);
  }
  #endif
  static auto op = c10::Dispatcher::singleton().findSchemaOrThrow("aten::__ilshift__", "Tensor");
  c10::Dispatcher::singleton().redispatch<Tensor &, Tensor &, const Tensor &>(op, c10::DispatchKey::Tracer, self, other);
  #if !defined(PYTORCH_DISABLE_TRACING)
  if (tracer_state) {
    jit::tracer::setTracingState(std::move(tracer_state));
    jit::tracer::addOutput(node, self);
  }
  #endif
  return self;
}
```



Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21570684/)!

Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] move tracing logic to a separate dispatch backend"


This PR moves tracing logic out of the generated VariableType kernels, to associate it with a new dedicated dispatch key Tracer.
It also toggles the dispatch key set at various places to keep the semantics unchanged - see the inline [Tracing Mode Switches] note.

Sample generated code:
```
Tensor & __ilshift___Tensor(Tensor & self, const Tensor & other) {
  #if !defined(PYTORCH_DISABLE_TRACING)
  torch::jit::Node* node = nullptr;
  std::shared_ptr<jit::tracer::TracingState> tracer_state;
  if (jit::tracer::isTracing()) {
    tracer_state = jit::tracer::getTracingState();
    at::Symbol op_name;
    op_name = jit::Symbol::fromQualString("aten::__ilshift__");
    node = tracer_state->graph->create(op_name, /*num_outputs=*/0);
    jit::tracer::recordSourceLocation(node);
    jit::tracer::addInputs(node, "self", self);
    jit::tracer::addInputs(node, "other", other);
    tracer_state->graph->insertNode(node);

    jit::tracer::setTracingState(nullptr);
  }
  #endif
  static auto op = c10::Dispatcher::singleton().findSchemaOrThrow("aten::__ilshift__", "Tensor");
  c10::Dispatcher::singleton().redispatch<Tensor &, Tensor &, const Tensor &>(op, c10::DispatchKey::Tracer, self, other);
  #if !defined(PYTORCH_DISABLE_TRACING)
  if (tracer_state) {
    jit::tracer::setTracingState(std::move(tracer_state));
    jit::tracer::addOutput(node, self);
  }
  #endif
  return self;
}
```



Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21570684/)!

Differential Revision: [D21570684](https://our.internmc.facebook.com/intern/diff/D21570684)

[ghstack-poisoned]
